8304831: TextFlow.hitTest.insertionIndex incorrect with surrogate pairs

karthikpandelu opened the pull request:
Since surrogate pairs are internally considered as 2 characters and text field is null in `HitInfo` when `getInsertionIndex` is invoked from `TextFlow`, wrong insertion index was returned.

Updated code to calculate insertion index in `getHitInfo` method of `PrismTextLayout` class when `hitTest` of trailing side of surrogate pair is requested. Since text runs are processed in this method already, calculating the insertion index in this method looks better than calculating in `getInsertionIndex` of `HitInfo` method.
The latter approach also requires the text to be sent to `HitInfo` as parameter from the `hitTest` method of `TextFlow`. If the number of `Text` nodes in `TextFlow` are very large, processing all the `Text` nodes on each `hitTest` method invocation might cause performance issue. Hence implemented first approach.

Added system test to validate the fix.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8304831](https://bugs.openjdk.org/browse/JDK-8304831): TextFlow.hitTest.insertionIndex incorrect with surrogate pairs (**Bug** - P3)


### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1091/head:pull/1091` \
`$ git checkout pull/1091`

Update a local copy of the PR: \
`$ git checkout pull/1091` \
`$ git pull https://git.openjdk.org/jfx.git pull/1091/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1091`

View PR using the GUI difftool: \
`$ git pr show -t 1091`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1091.diff">https://git.openjdk.org/jfx/pull/1091.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1091#issuecomment-1508071660)

kevinrushforth commented:
Reviewers: @prrace @andy-goryachev-oracle

andy-goryachev-oracle reviewed (commented):
could you please check with the Monkey Tester?
I could be doing something wrong.

karthikpandelu commented on tests/system/src/test/java/test/robot/javafx/scene/TextFlowSurrogatePairInsertionIndexTest.java:
```diff
@@ -0,0 +1,170 @@
+        Util.waitForLatch(mouseClickLatch, 10, "Timeout waiting for mouse click");
+    }
+
+    @Test
```
Sure. I will try to add these test cases.

andy-goryachev-oracle commented:
> I'm also seeing same issue

I wonder what could be the difference?  text size?

karthikpandelu commented:
> > I'm also seeing same issue
> 
> I wonder what could be the difference? text size?

Text size is one difference.
One more observation I made is that if the emojis and text in the TextFlow are separate text nodes then this fix works. If one text node contains combination of text and emojis then it is failing.
Also the fix does not work with the first emoji in the following picture but works with second one.
<img width="30" alt="image" src="https://user-images.githubusercontent.com/26969459/234771077-a5e59e05-c520-4b44-b0ff-0375f7832e48.png">

So looks like the the combination of different text and emojis are giving different results. I will check these and resume work on this PR after the test sprint.

karthikpandelu commented:
Fixed the above mentioned issue in calculating insertion index by adding insertion index calculation logic to `getHitInfo()` method. The previous fix was not calculating insertion index for emojis represented using value from 0x231A to 0x3299.
Added additional test cases as suggested above. 
Please review the changes.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java:
```diff
@@ -452,13 +452,23 @@ public Hit getHitInfo(float x, float y) {
-                    if (Character.isHighSurrogate(txt.charAt(relIdx)) && !leading) {
-                        insertionIndex = charIndex + 2;
+                insertionIndex = charIndex;
+                String txt = new String(getText());
```
possible NPE: looks like getText() might return null, in which case this line will throw an NPE.

karthikpandelu replied:
Updated code

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java:
```diff
@@ -452,13 +452,23 @@ public Hit getHitInfo(float x, float y) {
+                insertionIndex = charIndex;
+                String txt = new String(getText());
+                if (!leading) {
+                    if (txt != null) {
```
This check should be done earlier for getText()

karthikpandelu replied:
Added null check for `getText()`

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java:
```diff
@@ -452,13 +452,23 @@ public Hit getHitInfo(float x, float y) {
+                    if (txt != null) {
+                        int next;
+                        BreakIterator charIterator = BreakIterator.getCharacterInstance();
+                        synchronized(charIterator) {
```
question: why do we need to synchronize on the instance here?  BreakIterator.get*Instance() creates a new instance each time, so synchronization is probably unnecessary.

karthikpandelu replied:
Correct it is not necessary. Updated code to remove synchronize

andy-goryachev-oracle reviewed (changes requested):
minor changes requested

andy-goryachev-oracle commented:
Does not seem to work correctly still - the insertion index in TextFlow should not point in between surrogate pair of characters that comprise an emoji.  Basically, it should be exactly the same as with Text.hitInfo():

![Screenshot 2023-05-08 at 09 29 35](https://user-images.githubusercontent.com/107069028/236879369-f7634c6e-e5c8-41f4-8ded-dd59327de73b.png)

karthikpandelu commented:
> Does not seem to work correctly still - the insertion index in TextFlow should not point in between surrogate pair of characters that comprise an emoji. Basically, it should be exactly the same as with Text.hitInfo():
> 
This condition is working for me as `Text.hitInfo()`. The code is updated now to make the changes suggested above. Could you please recheck?

andy-goryachev-oracle reviewed (commented):
Testing looks good so far.

There is one question whether it's possible to always initialize HitInfo.insertionIndex and remove the secondary computation from HitInfo.getInsertionIndex() 

I am not clear under which conditions it is not possible to initialize, and whether subsequent HitInfo.getInsertionIndex()  would produce a meaningful result.  

Perhaps in a separate RFE.

karthikpandelu commented:
> There is one question whether it's possible to always initialize HitInfo.insertionIndex and remove the secondary computation from HitInfo.getInsertionIndex()
> 
> I am not clear under which conditions it is not possible to initialize, and whether subsequent HitInfo.getInsertionIndex() would produce a meaningful result.
> 
I think it should be possible to initialize insertionIndex always in `getHitInfo()` method. Only scenario now it is not getting initialized is when `lineIndex >= getLineCount()` is true and we don't check for text runs further and initialize `charIndex` using `getCharCount()` method.
I'm not completely clear if removing the `insertionIndex` calculation from `HitInfo.getInsertionIndex()` wouldn't cause any issue. So kept that as well.
Like you mentioned, this can be considered in a separate RFE I think.

karthikpandelu commented on modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java:
```diff
@@ -422,6 +422,7 @@ public PathElement[] getCaretShape(int offset, boolean isLeading,
     @Override
     public Hit getHitInfo(float x, float y) {
         int charIndex = -1;
+        int insertionIndex = -1;
```
After current fix, only scenario where `insertionIndex` not getting initialized will be when `lineIndex >= getLineCount()` is true. So even if `insertionIndex` -1, text will be null and `HitInfo.getInsertionIndex()` will not have any computation to perform. So we can resolve the insertion index always I think. Please let me know your thoughts on this.

andy-goryachev-oracle commented:
Just by looking at the code, the code paths that leave `insertionIndex=-1` in `PrismTextLayout.getHitInfo()` are on lines:
431, 450(when run==null), 472

If we can initialize insertionIndex to a positive value, then the buggy code in HitInfo.getInsertionIndex() will never get executed, and we can remove it later (in JDK-8302511).

karthikpandelu commented:
> Just by looking at the code, the code paths that leave `insertionIndex=-1` in `PrismTextLayout.getHitInfo()` are on lines: 431, 450(when run==null), 472
> 
> If we can initialize insertionIndex to a positive value, then the buggy code in HitInfo.getInsertionIndex() will never get executed, and we can remove it later (in JDK-8302511).

Updated the code to initialize `insertionIndex` in `HitInfo.getInsertionIndex()`. Please check.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java:
```diff
@@ -450,13 +451,35 @@ public Hit getHitInfo(float x, float y) {
+        if (insertionIndex == -1) {
+            insertionIndex = charIndex;
+            if (!leading) {
+                insertionIndex += 1;
```
I would have initialized insertionIndex directly in all the blocks instead of adding conditional logic, but this code does seem to produce the desired result. 
For clarify, would it be possible to avoid this additional conditional logic?

Also, do existing unit tests cover the newly modified paths?

karthikpandelu replied:
Initializing the insertion index directly with character index looks better. In this way we can avoid additional conditional logic as well. Changed code to do so.
Added additional test case. This tests the condition present in line 430.
I couldn't create the scenario to hit line 473. If you have any suggestions please let me know.

While testing I found that `testTextFlowInsertionIndexUsingMultipleEmojis()` was failing intermittently when all the tests are run. Hence added small delay.

andy-goryachev-oracle replied:
> I couldn't create the scenario to hit line 473. If you have any suggestions please let me know.

I can't seem to trigger this code path (I wonder if it's possible to hit it at all).  Anyway, the code there looks correct to me, even if this condition would never happen.

prrace replied:
Assuming line 473 is the one still there today, it looks to me as if that would be reached if you had the caret on an empty line that isn't the last line.

prrace replied:
Regarding
"If we can initialize insertionIndex to a positive value, then the buggy code in HitInfo.getInsertionIndex()
will never get executed, and we can remove it later (in JDK-8302511)."

it seems to me that this new code has been pretty much copied from the supposedly buggy code there .. 
I'm not sure that code is actually buggy (comments over in that bug), rather that the instance was constructed
with bad data.

andy-goryachev-oracle replied:
The main issue is that insertion index was not computed in the PrismTextLayout, supposedly to allow for delaying the computation until it's needed (if needed at all).  A good intention, but the execution had two issues:
1. HitInfo obtained in TextFlow had its 'text' field set to null, disabling any possibility of computing the index correctly
2. The code that computed the index had an issue in that if an exception happened, it left the object in an inconsistent state, leading to other problems down the road, and more exceptions (incl. AIOB)

What we did here is to compute the insertion index correctly at the source.
I've also checked that in the case of an empty line the insertion index is correct in both cases (at the last line and in the middle of a document).

karthikpandelu replied:
> Assuming line 473 is the one still there today, it looks to me as if that would be reached if you had the caret on an empty line that isn't the last line.

In this case also code hits the condition in 451 as the text run is not null.

>it seems to me that this new code has been pretty much copied from the supposedly buggy code there ..
I'm not sure that code is actually buggy (comments over in that bug), rather that the instance was constructed
with bad data.

Like Andy said, since text was null in `HitInfo::getInsertionIndex()` method in the case of TextFlow, correct insertion index was not calculated. I had to use pretty much same code in PrismTextLayout as I had to consider all the cases especially the U+FE0F Variation Selector-16 characters.

prrace replied:
I did note that TextFlow passed null which really looked like a WIP to me but its a separate issue.
The exception in 8302511 which is the entire subject of that bug isn't possible if text == null.
My point is that if you pass an incorrect charIndex to BreakIterator.following() you'll get the same exception
in your copy of that code. So how are you making sure you never do that ?

andy-goryachev-oracle replied:
The exception in 8302511 was likely caused by a prior exception due to null text, which corrupted insertionIndex value in HitInfo.

The goal of this change is to always compute insertionIndex (an I believe we do it correctly this time).

prrace replied:
No, it wasn't due to null text. That is provably impossible.

The exception from JDK-8302511 is as follows (copied from the bug report).

Exception in thread "JavaFX Application Thread" java.lang.IllegalArgumentException: offset out of bounds 
at java.base/sun.text.RuleBasedBreakIterator.checkOffset(RuleBasedBreakIterator.java:730) 
at java.base/sun.text.RuleBasedBreakIterator.following(RuleBasedBreakIterator.java:744) 
at javafx.graphics/javafx.scene.text.HitInfo.getInsertionIndex(HitInfo.java:84) 
at javafx.graphics/javafx.scene.text.HitInfo.toString(HitInfo.java:100) 
Look at the code in HitInfo.java referenced in JDK-8302511,  the code at line 84 can only be reached if text != null
..
              if (text != null) {
                    // Skip complex character clusters / ligatures.
                    int next;
                    synchronized(charIterator) {
                        charIterator.setText(text);
                        next = charIterator.following(insertionIndex);   // this is line 84
                    }

...

andy-goryachev-oracle replied:
you are right, it cannot be explained by a null text.

so if i understand you correctly, the question is about validity of `insertionIndex += 1` on lines 463 and 469.

463: should it be set to text.length() instead?

469: since text==null, could `insertionIndex = charIndex + 1` be correct?  

And also, is it possible at all that text == null within PrismTextLayout?

prrace replied:
No, because that's after the call to  BreakIterator.following(int)
It is mainly about the assignment on line 456 
> insertionIndex = charIndex;

since that is what gets passed to BreakIterator.following(int)

and charIndex was calculated by
charIndex = run.getStart() + run.getOffsetAtX(x, trailing);
so it goes back to that calculation.
Perhaps a simple validation that charIndex is not out of bounds is all that is needed.

karthikpandelu replied:
Added check to make sure that insertion index passed to BreakIterator.following(int) is not out of bound.
Added this check after line 456 `insertionIndex = charIndex;`, otherwise insertionIndex remains as -1 and  `Hitinfo::getInsertionIndex` will calculate wrong insertionIndex.
Please check the change.

andy-goryachev-oracle commented:
Karthik, we are very close!
I see one consistent failure on macOS 13.3.1(a):
```
java.lang.AssertionError: expected:<0> but was:<-2>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at test.robot.javafx.scene.TextFlowSurrogatePairInsertionIndexTest.testTextFlowInsertionIndexUsingMultipleEmojis(TextFlowSurrogatePairInsertionIndexTest.java:227)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

Also, I noticed that the tests do not cover a case of line wrapping - would it be possible to add a similar test where the text string is wrapped, so we can verify that monotonic test works across the wrap point please?

karthikpandelu commented:
Sure. I will check if I can add new test case with wrapped text.
I was also getting same failure but after the latest fix I didn't find failure in `testTextFlowInsertionIndexUsingMultipleEmojis` test. I will check this again.

karthikpandelu commented:
Added new test to validate character index and insertion index in wrapped text.
Made optimizations to use single mouse move function in all the tests.

@andy-goryachev-oracle , above failure is caused because test is asserting before the mouse event handler could retrive hitinfo values. So added small delays in all the tests to make sure that the text is rendered and test becomes stable before getting the hitinfo values and assertion. Please let me know if you are able to run all the tests.

karthikpandelu commented:
This looks like timing issue only. I believe we have to increase the delays little bit more.

andy-goryachev-oracle commented:
OK, I think I know what the problem is.
Instead of adding `Thread.sleep(500);`, we should add `Toolkit.getToolkit().firePulse();` before `textSetLatch.countDown();` (after we add all children to TextFlow).

andy-goryachev-oracle commented on tests/system/src/test/java/test/robot/javafx/scene/TextFlowSurrogatePairInsertionIndexTest.java:
```diff
@@ -0,0 +1,381 @@
+            text.setFont(new Font(48));
+            textFlow.getChildren().clear();
+            textFlow.getChildren().setAll(text);
+
```
let's add Toolkit.getToolkit().firePulse(); before textSetLatch.countDown();

andy-goryachev-oracle commented on tests/system/src/test/java/test/robot/javafx/scene/TextFlowSurrogatePairInsertionIndexTest.java:
```diff
@@ -0,0 +1,381 @@
+    @Test
+    public void testTextFlowInsertionIndexUsingWrappedText() throws Exception {
+        addLongText();
+        Thread.sleep(500);
```
we can remove Thread.sleep() now

andy-goryachev-oracle reviewed (changes requested):
please try adding firePulse(), and also please check on Windows and Linux.
thanks!

karthikpandelu commented:
As I understood from another [PR discussion](https://github.com/openjdk/jfx/pull/1119#discussion_r1180380110), `Toolkit.getToolkit().firePulse();` will not make any difference in system tests. I think it is required only for headless unit tests.

andy-goryachev-oracle commented:
> `Toolkit.getToolkit().firePulse();` will not make any difference in system tests.

Once you added children to TextFlow, I think it needs another layout pass to apply CSS and render the text.  We could use delays, but it looks like sometimes even 500 ms delay is not sufficient (possibly due to cold start), making the test unstable.

I think I've also encountered a few other tests where explicit firePulse() was needed to allow for skin, CSS, and layout pass.

kevinrushforth commented:
I'd be interested to know whether the call to `Toolkit::firePulse` makes this test stable. A system test really should not need to call `firePulse` directly. It artificially alters the control flow.

Perhaps a better approach, which we could consider for a follow-up, would be to have a utility method that adds a pulse listener, calls `Platform::requestNextPulse`, and waits on a latch triggered by the pulse listener.

karthikpandelu commented:
Adding `Toolkit::firePulse` is not making the tests stable. In fact I'm getting new failure after adding this in `testTextFlowInsertionIndexUsingMultipleEmojis()` test.
```
TextFlowSurrogatePairInsertionIndexTest > testTextFlowInsertionIndexUsingMultipleEmojis FAILED
    java.lang.NullPointerException: Cannot read the array length because "runs" is null
        at javafx.graphics@21-internal/javafx.scene.text.Text.getSpanBounds(Text.java:359)
        at javafx.graphics@21-internal/javafx.scene.text.Text.doComputeGeomBounds(Text.java:1159)
        at javafx.graphics@21-internal/javafx.scene.text.Text$1.doComputeGeomBounds(Text.java:149)
        at javafx.graphics@21-internal/com.sun.javafx.scene.shape.TextHelper.computeGeomBoundsImpl(TextHelper.java:90)
        at javafx.graphics@21-internal/com.sun.javafx.scene.NodeHelper.computeGeomBounds(NodeHelper.java:117)
        at javafx.graphics@21-internal/javafx.scene.Node.updateGeomBounds(Node.java:3813)
        at javafx.graphics@21-internal/javafx.scene.Node.getGeomBounds(Node.java:3775)
        at javafx.graphics@21-internal/javafx.scene.Node.getLocalBounds(Node.java:3723)
        at javafx.graphics@21-internal/javafx.scene.Node.updateTxBounds(Node.java:3877)
        at javafx.graphics@21-internal/javafx.scene.Node.getTransformedBounds(Node.java:3669)
        at javafx.graphics@21-internal/javafx.scene.Node.updateBounds(Node.java:777)
        at javafx.graphics@21-internal/javafx.scene.Parent.updateBounds(Parent.java:1834)
        at javafx.graphics@21-internal/javafx.scene.Scene$ScenePulseListener.pulse(Scene.java:2615)
        at javafx.graphics@21-internal/com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Toolkit.java:401)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at javafx.graphics@21-internal/com.sun.javafx.tk.Toolkit.runPulse(Toolkit.java:400)
        at javafx.graphics@21-internal/com.sun.javafx.tk.Toolkit.firePulse(Toolkit.java:430)
        at test.robot.javafx.scene.TextFlowSurrogatePairInsertionIndexTest.testTextFlowInsertionIndexUsingMultipleEmojis(TextFlowSurrogatePairInsertionIndexTest.java:233)
```
Could increasing the delay time is the only solution now for the test failure seen previously and then we follow up with the approach suggested by @kevinrushforth ?

@andy-goryachev-oracle  could you please help me in understanding how `Util.runAndWait(() -> { });` help in this case? 
Also I'm using the `Util.runAndWait(() -> { });` while setting text to textflow and trigger `textSetLatch.countDown()`. Once `textSetLatch` count down to 0, then I proceed with the test.

andy-goryachev-oracle commented:
> value 10084

This character is a U+FE0F Variation Selector-16 
https://codepoints.net/U+FE0F?lang=en
```
This codepoint may change the appearance of the preceding character. If that is a symbol, dingbat or emoji, U+FE0F forces it to be rendered as a colorful image as compared to a monochrome text variant.
```

I suspect the expected behavior is not supported by JavaFX along with emoji modifier sequences (@prrace please confirm) - I see TextArea and TextFlowSurrogatePairInsertionIndexTest both render "\u2764\ufe0f" sequence as a back heart, while Eclipse renders it correctly as a red heart.

For the purposes of this test, we could remove this sequence from the test, I think.

andy-goryachev-oracle commented:
I don't mind someone else take a look at this, @kevinrushforth .

kevinrushforth commented:
@prrace Would you be willing to take a  look?

/reviewers 2

andy-goryachev-oracle commented:
what do you people think of the number of pulses parameter?  is 10 enough?  is there a reasonable upper limit that's needed for any processing to settle?

another question is whether it could be made static - i.e. not require a Scene.  Perhaps we could add a pulse listener to all visible Scenes?  Would that be sufficient, since we only need a signal from one to indicate the pulse has been performed, and there is another and another (... up to 10 or N) pulses that basically guarantees that things are settled.

And if the layout has not settled by that time, we have a bigger issue (can we detect this?  I've seen continuous layout loops in one of my programs)

kevinrushforth commented:
> what do you people think of the number of pulses parameter? is 10 enough? is there a reasonable upper limit that's needed for any processing to settle?

I think 10 should be a reasonable number of pulses for our tests. Most graphs settle down within 3 layout passes.

> another question is whether it could be made static - i.e. not require a Scene. Perhaps we could add a pulse listener to all visible Scenes? Would that be sufficient, since we only need a signal from one to indicate the pulse has been performed, and there is another and another (... up to 10 or N) pulses that basically guarantees that things are settled.

I doubt this is worth it. This is a test utility, and if one of the very few tests that have multiple scenes needed to call it, it wouldn't be hard to have the test call it for each scene (and would give the test more control).

> And if the layout has not settled by that time, we have a bigger issue (can we detect this? I've seen continuous layout loops in one of my programs)

Right. This is tangential to this PR, but at some point we might want to resurrect the discussion on PR #433 to see if we can improve the current situation.

karthikpandelu commented:
> Can you file a new (P4) testbug as a follow-up issue to use it replace the two existing uses of `firePulse`?

Created [JDK-8308608](https://bugs.openjdk.org/browse/JDK-8308608)

andy-goryachev-oracle commented:
The new code also exhibits a strange behavior in the RichTextArea (based ob TextFlow): I am unable to position the cursor at the end of ☝🏿☝🏿☝🏿 emoji sequence with the mouse (unless I click on the next empty line).

karthikpandelu commented:
Thanks @andy-goryachev-oracle. I will check above mentioned issue.

karthikpandelu commented on tests/system/src/test/java/test/robot/javafx/scene/TextFlowSurrogatePairInsertionIndexTest.java:
```diff
@@ -0,0 +1,355 @@
+
+    private void addMultipleEmojis() {
+        Util.runAndWait(() -> {
+            text = new Text("😊😇💙🦋🏁🔥");
```
This case is same as the one which we saw with red heart as the emoji sequence mentioned here is also a U+FE0F Variation Selector-16 character. Since it exhibits different behavior in different platforms we decided to remove it.

I compared the insertion index calculated for these emojis with the insertion index calculated for red heart emoji. The insertion index calculated looks to be correct for the above given emojis. It is considered as 2 characters if color is also present else it is considered as single character.
For me it is giving correct caret position as well in the Monkey tester.

I found strange behavior when I tried to add heart emojis (with and without color) along with the `CLUSTERS` variable present in Monkey Tester so that it gets displayed when Rich Text option is selected.
When `CLUSTERS = "❤️❤❤️"`, emojis are displayed properly.
<img width="658" alt="image" src="https://github.com/openjdk/jfx/assets/26969459/f3de5d1d-ad8c-42a6-8284-01916dfd37a8">

When `CLUSTERS = "☝🏿☝🏿☝🏿🤦🏼‍♂️❤️❤❤️";`, following sequence is displayed.
<img width="532" alt="image" src="https://github.com/openjdk/jfx/assets/26969459/6eaa67b8-6fa1-475f-82bd-e9c48240b1c1">

This looks like a separate issue. Please let me know your thoughts on this.

andy-goryachev-oracle replied:
@karthikpandelu : 
1. you are right, multiple symbols and a failed test are due to lack of proper handling of grapheme clusters in fx.  I don't think we have a ticket, so created JDK-8309565 for that
2. your latest fix uncovered a problem in my RichTextArea code, so thank you!

The new code passes all my tests in the MonkeyTester and RichTextArea.

karthikpandelu replied:
Than you @andy-goryachev-oracle for the detailed review and feedback.

karthikpandelu commented:
/integrate
